File: src/test/java/no/nav/syfo/service/SaksbehandlingsServiceTest.java
package no.nav.syfo.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;
import no.nav.syfo.consumer.repository.InnsendingDAO;
import no.nav.syfo.consumer.ws.*;
import no.nav.syfo.domain.Soknad;
import no.nav.syfo.domain.dto.Sykepengesoknad;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.IOException;

import static no.nav.syfo.TestUtils.soknadSelvstendigMedNeisvar;
import static no.nav.syfo.domain.dto.Soknadstype.SELVSTENDIGE_OG_FRILANSERE;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class SaksbehandlingsServiceTest {

    @Mock
    private AktorConsumer aktorConsumer;
    @Mock
    private PersonConsumer personConsumer;
    @Mock
    private InnsendingDAO innsendingDAO;
    @Mock
    private BehandleSakConsumer behandleSakConsumer;
    @Mock
    private BehandleJournalConsumer behandleJournalConsumer;
    @Mock
    private BehandlendeEnhetService behandlendeEnhetService;
    @Mock
    private OppgavebehandlingConsumer oppgavebehandlingConsumer;
    @Mock
    private MeterRegistry registry;

    @InjectMocks
    private SaksbehandlingsService saksbehandlingsService;

    private ObjectMapper objectMapper = new ObjectMapper().registerModule(new JavaTimeModule());

    @Before
    public void setup() {
        when(aktorConsumer.finnFnr(any())).thenReturn("12345678901");
        when(personConsumer.finnBrukerPersonnavnByFnr(any())).thenReturn("Personnavn");
        when(behandleSakConsumer.opprettSak(any())).thenReturn("saksId");
        when(behandleJournalConsumer.opprettJournalpost(any(), any())).thenReturn("journalpostId");
        when(behandlendeEnhetService.hentBehandlendeEnhet("12345678901", SELVSTENDIGE_OG_FRILANSERE)).thenReturn("2017");
        when(oppgavebehandlingConsumer.opprettOppgave(anyString(), anyString(), anyString(), anyString(), any(Soknad.class))).thenReturn("oppgaveId");
        when(registry.counter(any(), anyIterable())).thenReturn(mock(Counter.class));
    }

    @Test
    public void behandlerInnsending() throws IOException {
        Sykepengesoknad sykepengesoknad = objectMapper.readValue(soknadSelvstendigMedNeisvar, Sykepengesoknad.class);

        saksbehandlingsService.behandleSoknad(sykepengesoknad);

        verify(innsendingDAO).settBehandlet(any());
    }

    @Test
    public void feilendeInnsendingLeggesIBasen() throws IOException {
        when(behandleJournalConsumer.opprettJournalpost(any(), any()))
                .thenThrow(new RuntimeException("Opprett journal feilet"));

        Sykepengesoknad sykepengesoknad = objectMapper.readValue(soknadSelvstendigMedNeisvar, Sykepengesoknad.class);
        saksbehandlingsService.behandleSoknad(sykepengesoknad);

        verify(innsendingDAO, times(1)).leggTilFeiletInnsending(any());
    }
}

File: src/test/kotlin/no/nav/syfo/service/SaksbehandlingsServiceTest.kt
package no.nav.syfo.service

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.argumentCaptor
import io.micrometer.core.instrument.Counter
import io.micrometer.core.instrument.MeterRegistry
import no.nav.syfo.any
import no.nav.syfo.arkivering.Arkivaren
import no.nav.syfo.client.FlexBucketUploaderClient
import no.nav.syfo.client.pdl.PdlClient
import no.nav.syfo.domain.dto.Sykepengesoknad
import no.nav.syfo.kafka.producer.RebehandleSykepengesoknadProducer
import no.nav.syfo.repository.InnsendingDbRecord
import no.nav.syfo.repository.InnsendingRepository
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.junit.jupiter.MockitoSettings
import org.mockito.quality.Strictness
import java.time.Instant
import java.time.LocalDateTime
import java.util.UUID

@ExtendWith(MockitoExtension::class)
@MockitoSettings(strictness = Strictness.LENIENT)
class SaksbehandlingsServiceTest {
    @Mock
    lateinit var pdlClient: PdlClient
    @Mock
    lateinit var identService: IdentService
    @Mock
    lateinit var innsendingRepository: InnsendingRepository
    @Mock
    lateinit var oppgaveService: OppgaveService
    @Mock
    lateinit var arkivaren: Arkivaren
    @Mock
    lateinit var flexBucketUploaderClient: FlexBucketUploaderClient
    @Mock
    lateinit var registry: MeterRegistry
    @Mock
    lateinit var rebehandleSykepengesoknadProducer: RebehandleSykepengesoknadProducer

    @InjectMocks
    lateinit var saksbehandlingsService: SaksbehandlingsService

    private val objectMapper = ObjectMapper().registerKotlinModule().registerModules(JavaTimeModule())
    private val aktorId = "aktorId-745463060"
    private val fnr = "12345678901"

    @BeforeEach
    fun setup() {
        given(pdlClient.hentFormattertNavn(any())).willReturn("Personnavn")
        given(identService.hentAktorIdForFnr(any())).willReturn(aktorId)
        given(identService.hentFnrForAktorId(any())).willReturn(fnr)
        given(arkivaren.opprettJournalpost(any())).willReturn("journalpostId")
        given(oppgaveService.opprettOppgave(any())).willReturn(OppgaveResponse(1234))
        given(
            registry.counter(
                ArgumentMatchers.anyString(),
                ArgumentMatchers.anyIterable()
            )
        ).willReturn(mock(Counter::class.java))
        given(innsendingRepository.save(any())).willReturn(
            InnsendingDbRecord(
                id = UUID.randomUUID().toString(),
                sykepengesoknadId = "innsending-guid"
            )
        )
    }

    @Test
    fun behandlerSoknaderSomEttersendesTilNavDerDetManglerOppgave() {
        val now = LocalDateTime.now()
        val sykepengesoknadUtenOppgave = objectMapper.readValue(SaksbehandlingsServiceTest::class.java.getResource("/soknadArbeidstakerMedNeisvar.json"), Sykepengesoknad::class.java)
            .copy(sendtNav = null, sendtArbeidsgiver = now)
        val sykepengesoknadEttersendingTilNAV = objectMapper.readValue(SaksbehandlingsServiceTest::class.java.getResource("/soknadArbeidstakerMedNeisvar.json"), Sykepengesoknad::class.java)
            .copy(sendtNav = now, sendtArbeidsgiver = now)

        saksbehandlingsService.behandleSoknad(sykepengesoknadUtenOppgave)
        verify(arkivaren, times(1)).opprettJournalpost(any())
        verify(oppgaveService, never()).opprettOppgave(any())
        given(innsendingRepository.findBySykepengesoknadId(sykepengesoknadUtenOppgave.id)).willReturn(
            innsending(sykepengesoknadUtenOppgave)
        )

        saksbehandlingsService.behandleSoknad(sykepengesoknadEttersendingTilNAV)
        saksbehandlingsService.opprettOppgave(sykepengesoknadEttersendingTilNAV, innsending(sykepengesoknadEttersendingTilNAV))
        verify(arkivaren, times(1)).opprettJournalpost(any())
        verify(oppgaveService, times(1)).opprettOppgave(any())
    }

    private fun innsending(
        sykepengesoknadUtenOppgave: Sykepengesoknad
    ): InnsendingDbRecord {
        return InnsendingDbRecord(
            id = "innsending-guid",
            sykepengesoknadId = sykepengesoknadUtenOppgave.id,
            journalpostId = "journalpostId",
            oppgaveId = null,
            behandlet = Instant.now()
        )
    }

    @Test
    fun `oppretter oppgave med korrekte felter`() {
        val captor: KArgumentCaptor<OppgaveRequest> = argumentCaptor()
        val søknad = objectMapper.readValue(SaksbehandlingsServiceTest::class.java.getResource("/soknadArbeidstakerMedNeisvar.json"), Sykepengesoknad::class.java)
        saksbehandlingsService.opprettOppgave(søknad, innsending(søknad))
        verify(oppgaveService).opprettOppgave(captor.capture())
        val oppgaveRequest = captor.firstValue
        Assertions.assertThat(oppgaveRequest.journalpostId).isEqualTo("journalpostId")
        Assertions.assertThat(oppgaveRequest.beskrivelse).isEqualTo(
            """
Søknad om sykepenger for perioden 16.10.2018 - 24.10.2018

Arbeidsgiver: ARBEIDSGIVER A/S
Organisasjonsnummer: 1257358

Periode 1:
16.10.2018 - 20.10.2018
Grad: 100

Periode 2:
21.10.2018 - 24.10.2018
Grad: 40

Betaler arbeidsgiveren lønnen din når du er syk?
Vet ikke
            """.trimIndent()
        )
        Assertions.assertThat(oppgaveRequest.tema).isEqualTo("SYK")
        Assertions.assertThat(oppgaveRequest.oppgavetype).isEqualTo("SOK")
        Assertions.assertThat(oppgaveRequest.prioritet).isEqualTo("NORM")
        Assertions.assertThat(oppgaveRequest.behandlingstema).isEqualTo("ab0061")
    }
}
